File: src/main/java/hibiki/ui/Theme.java
package hibiki.ui;

import java.awt.Color;
import java.awt.Font;
import java.util.ArrayList;
import java.util.List;

public class Theme {
  public enum Preset {
    ABLETON_DARK, ABLETON_LIGHT, SOLARIZED_DARK, SOLARIZED_LIGHT, WIN95, WINXP
  }

  private static Theme instance = new Theme(Preset.ABLETON_DARK, 1.0f, 11);

    // Dynamic properties
    public Color BG_DARKER;
    public Color BG_DARK;
    public Color BG_MEDIUM;
    public Color PANEL_BG;
    public Color PANEL_BG_LIGHT;
    public Color BORDER;
    public Color TEXT_NORMAL;
    public Color TEXT_BRIGHT;
    public Color TEXT_LIGHT;
    public Color TEXT_DIM;
    public Color ACCENT_ORANGE;
    public Color ACCENT_BLUE;
    public Color ACCENT_GREEN;
    public Color TRACK_HEADER;
    public Color CLIP_MIDI;
    public Color CLIP_AUDIO;
    public Color CLIP_PLAYING;

    public Font FONT_UI;
    public Font FONT_UI_BOLD;
    public Font FONT_DISPLAY;

    private float scaling = 1.0f;
    private int baseFontSize = 11;
    private Preset currentPreset = Preset.ABLETON_DARK;

    public interface ThemeListener {
      void onThemeChanged();
    }

    private final List<ThemeListener> listeners = new ArrayList<>();

    public static Theme getInstance() {
      return instance;
    }

    private Theme(Preset preset, float scaling, int fontSize) {
      update(preset, scaling, fontSize);
    }

    public void update(Preset preset, float scaling, int fontSize) {
      this.currentPreset = preset;
      this.scaling = scaling;
      this.baseFontSize = fontSize;

      applyPreset(preset);
      applyScaling();

      for (ThemeListener l : listeners) {
        l.onThemeChanged();
      }
    }

    public void addListener(ThemeListener l) {
      listeners.add(l);
    }

    public Preset getCurrentPreset() {
      return currentPreset;
    }

    public float getScaling() {
      return scaling;
    }

    public int getBaseFontSize() {
      return baseFontSize;
    }

    private void applyPreset(Preset preset) {
      switch (preset) {
        case ABLETON_DARK:
          BG_DARKER = new Color(25, 25, 25);
          BG_DARK = new Color(34, 34, 34);
          BG_MEDIUM = new Color(45, 45, 45);
          PANEL_BG = new Color(45, 45, 45);
          PANEL_BG_LIGHT = new Color(55, 55, 55);
          BORDER = new Color(20, 20, 20);
          TEXT_NORMAL = new Color(190, 190, 190);
          TEXT_BRIGHT = new Color(240, 240, 240);
          TEXT_LIGHT = new Color(220, 220, 220);
          TEXT_DIM = new Color(120, 120, 120);
          ACCENT_ORANGE = new Color(255, 153, 0);
          ACCENT_BLUE = new Color(0, 170, 255);
          ACCENT_GREEN = new Color(0, 255, 127);
          TRACK_HEADER = new Color(60, 60, 60);
          CLIP_MIDI = new Color(255, 120, 120);
          CLIP_AUDIO = new Color(120, 200, 255);
          CLIP_PLAYING = new Color(120, 255, 120);
          break;
        case ABLETON_LIGHT:
          BG_DARKER = new Color(220, 220, 220);
          BG_DARK = new Color(200, 200, 200);
          BG_MEDIUM = new Color(180, 180, 180);
          PANEL_BG = new Color(210, 210, 210);
          PANEL_BG_LIGHT = new Color(230, 230, 230);
          BORDER = new Color(160, 160, 160);
          TEXT_NORMAL = new Color(40, 40, 40);
          TEXT_BRIGHT = new Color(0, 0, 0);
          TEXT_LIGHT = new Color(60, 60, 60);
          TEXT_DIM = new Color(100, 100, 100);
          ACCENT_ORANGE = new Color(255, 153, 0);
          ACCENT_BLUE = new Color(0, 120, 255);
          ACCENT_GREEN = new Color(0, 200, 100);
          TRACK_HEADER = new Color(170, 170, 170);
          CLIP_MIDI = new Color(255, 120, 120);
          CLIP_AUDIO = new Color(120, 200, 255);
          CLIP_PLAYING = new Color(120, 255, 120);
          break;
        case SOLARIZED_DARK:
          BG_DARKER = new Color(0, 25, 30);
          BG_DARK = new Color(7, 54, 66);
          BG_MEDIUM = new Color(88, 110, 117);
          PANEL_BG = new Color(7, 54, 66);
          PANEL_BG_LIGHT = new Color(63, 85, 93);
          BORDER = new Color(0, 43, 54);
          TEXT_NORMAL = new Color(147, 161, 161);
          TEXT_BRIGHT = new Color(253, 246, 227);
          TEXT_LIGHT = new Color(131, 148, 150);
          TEXT_DIM = new Color(88, 110, 117);
          ACCENT_ORANGE = new Color(203, 75, 22);
          ACCENT_BLUE = new Color(38, 139, 210);
          ACCENT_GREEN = new Color(133, 153, 0);
          TRACK_HEADER = new Color(0, 43, 54);
          CLIP_MIDI = new Color(211, 54, 130);
          CLIP_AUDIO = new Color(38, 139, 210);
          CLIP_PLAYING = new Color(133, 153, 0);
          break;
        case SOLARIZED_LIGHT:
          BG_DARKER = new Color(240, 235, 215);
          BG_DARK = new Color(253, 246, 227);
          BG_MEDIUM = new Color(147, 161, 161);
          PANEL_BG = new Color(253, 246, 227);
          PANEL_BG_LIGHT = new Color(238, 232, 213);
          BORDER = new Color(211, 201, 171);
          TEXT_NORMAL = new Color(101, 123, 131);
          TEXT_BRIGHT = new Color(7, 54, 66);
          TEXT_LIGHT = new Color(88, 110, 117);
          TEXT_DIM = new Color(147, 161, 161);
          ACCENT_ORANGE = new Color(203, 75, 22);
          ACCENT_BLUE = new Color(38, 139, 210);
          ACCENT_GREEN = new Color(133, 153, 0);
          TRACK_HEADER = new Color(238, 232, 213);
          CLIP_MIDI = new Color(211, 54, 130);
          CLIP_AUDIO = new Color(38, 139, 210);
          CLIP_PLAYING = new Color(133, 153, 0);
          break;
        case WIN95:
          BG_DARKER = new Color(128, 128, 128);
          BG_DARK = new Color(192, 192, 192);
          BG_MEDIUM = new Color(223, 223, 223);
          PANEL_BG = new Color(192, 192, 192);
          PANEL_BG_LIGHT = new Color(255, 255, 255);
          BORDER = new Color(0, 0, 0);
          TEXT_NORMAL = new Color(0, 0, 0);
          TEXT_BRIGHT = new Color(0, 0, 0);
          TEXT_LIGHT = new Color(64, 64, 64);
          TEXT_DIM = new Color(128, 128, 128);
          ACCENT_ORANGE = new Color(0, 0, 128); // Title bar blue
          ACCENT_BLUE = new Color(0, 0, 255);
          ACCENT_GREEN = new Color(0, 128, 0);
          TRACK_HEADER = new Color(128, 128, 128);
          CLIP_MIDI = new Color(255, 0, 255);
          CLIP_AUDIO = new Color(0, 255, 255);
          CLIP_PLAYING = new Color(0, 255, 0);
          break;
        case WINXP:
          BG_DARKER = new Color(212, 208, 200);
          BG_DARK = new Color(236, 233, 216);
          BG_MEDIUM = new Color(255, 255, 255);
          PANEL_BG = new Color(236, 233, 216);
          PANEL_BG_LIGHT = new Color(255, 255, 255);
          BORDER = new Color(172, 168, 153);
          TEXT_NORMAL = new Color(0, 0, 0);
          TEXT_BRIGHT = new Color(0, 0, 0);
          TEXT_LIGHT = new Color(80, 80, 80);
          TEXT_DIM = new Color(150, 150, 150);
          ACCENT_ORANGE = new Color(0, 85, 225); // XP Blue
          ACCENT_BLUE = new Color(49, 106, 197);
          ACCENT_GREEN = new Color(58, 110, 165);
          TRACK_HEADER = new Color(192, 192, 192);
          CLIP_MIDI = new Color(255, 100, 100);
          CLIP_AUDIO = new Color(100, 100, 255);
          CLIP_PLAYING = new Color(100, 255, 100);
          break;
      }
    }

    private void applyScaling() {
      int scaledSize = (int) (baseFontSize * scaling);
      FONT_UI = new Font("SansSerif", Font.PLAIN, scaledSize);
      FONT_UI_BOLD = new Font("SansSerif", Font.BOLD, scaledSize);
      FONT_DISPLAY = new Font("Monospaced", Font.BOLD, (int) (14 * scaling));
    }

    // Helper for scaling dimensions
    public int scale(int val) {
      return (int) (val * scaling);
    }

    public float scale(float val) {
      return val * scaling;
    }
}

File: src/main/clojure/hibiki/ui/theme.clj
(ns hibiki.ui.theme
  "DAW theme system — colors, fonts, scaling, and presets.
   Ported from Theme.java."
  (:import [java.awt Color Font]))

;; ---------------------------------------------------------------------------
;; Presets — each is a map of keyword -> Color
;; ---------------------------------------------------------------------------

(def presets
  {:ableton-dark
   {:bg-darker      (Color. 25 25 25)
    :bg-dark        (Color. 34 34 34)
    :bg-medium      (Color. 45 45 45)
    :panel-bg       (Color. 45 45 45)
    :panel-bg-light (Color. 55 55 55)
    :border         (Color. 20 20 20)
    :text-normal    (Color. 190 190 190)
    :text-bright    (Color. 240 240 240)
    :text-light     (Color. 220 220 220)
    :text-dim       (Color. 120 120 120)
    :accent-orange  (Color. 255 153 0)
    :accent-blue    (Color. 0 170 255)
    :accent-green   (Color. 0 255 127)
    :track-header   (Color. 60 60 60)
    :clip-midi      (Color. 255 120 120)
    :clip-audio     (Color. 120 200 255)
    :clip-playing   (Color. 120 255 120)}

   :ableton-light
   {:bg-darker      (Color. 220 220 220)
    :bg-dark        (Color. 200 200 200)
    :bg-medium      (Color. 180 180 180)
    :panel-bg       (Color. 210 210 210)
    :panel-bg-light (Color. 230 230 230)
    :border         (Color. 160 160 160)
    :text-normal    (Color. 40 40 40)
    :text-bright    (Color. 0 0 0)
    :text-light     (Color. 60 60 60)
    :text-dim       (Color. 100 100 100)
    :accent-orange  (Color. 255 153 0)
    :accent-blue    (Color. 0 120 255)
    :accent-green   (Color. 0 200 100)
    :track-header   (Color. 170 170 170)
    :clip-midi      (Color. 255 120 120)
    :clip-audio     (Color. 120 200 255)
    :clip-playing   (Color. 120 255 120)}

   :solarized-dark
   {:bg-darker      (Color. 0 25 30)
    :bg-dark        (Color. 7 54 66)
    :bg-medium      (Color. 88 110 117)
    :panel-bg       (Color. 7 54 66)
    :panel-bg-light (Color. 63 85 93)
    :border         (Color. 0 43 54)
    :text-normal    (Color. 147 161 161)
    :text-bright    (Color. 253 246 227)
    :text-light     (Color. 131 148 150)
    :text-dim       (Color. 88 110 117)
    :accent-orange  (Color. 203 75 22)
    :accent-blue    (Color. 38 139 210)
    :accent-green   (Color. 133 153 0)
    :track-header   (Color. 0 43 54)
    :clip-midi      (Color. 211 54 130)
    :clip-audio     (Color. 38 139 210)
    :clip-playing   (Color. 133 153 0)}

   :solarized-light
   {:bg-darker      (Color. 240 235 215)
    :bg-dark        (Color. 253 246 227)
    :bg-medium      (Color. 147 161 161)
    :panel-bg       (Color. 253 246 227)
    :panel-bg-light (Color. 238 232 213)
    :border         (Color. 211 201 171)
    :text-normal    (Color. 101 123 131)
    :text-bright    (Color. 7 54 66)
    :text-light     (Color. 88 110 117)
    :text-dim       (Color. 147 161 161)
    :accent-orange  (Color. 203 75 22)
    :accent-blue    (Color. 38 139 210)
    :accent-green   (Color. 133 153 0)
    :track-header   (Color. 238 232 213)
    :clip-midi      (Color. 211 54 130)
    :clip-audio     (Color. 38 139 210)
    :clip-playing   (Color. 133 153 0)}

   :win95
   {:bg-darker      (Color. 128 128 128)
    :bg-dark        (Color. 192 192 192)
    :bg-medium      (Color. 223 223 223)
    :panel-bg       (Color. 192 192 192)
    :panel-bg-light (Color. 255 255 255)
    :border         (Color. 0 0 0)
    :text-normal    (Color. 0 0 0)
    :text-bright    (Color. 0 0 0)
    :text-light     (Color. 64 64 64)
    :text-dim       (Color. 128 128 128)
    :accent-orange  (Color. 0 0 128)
    :accent-blue    (Color. 0 0 255)
    :accent-green   (Color. 0 128 0)
    :track-header   (Color. 128 128 128)
    :clip-midi      (Color. 255 0 255)
    :clip-audio     (Color. 0 255 255)
    :clip-playing   (Color. 0 255 0)}

   :winxp
   {:bg-darker      (Color. 212 208 200)
    :bg-dark        (Color. 236 233 216)
    :bg-medium      (Color. 255 255 255)
    :panel-bg       (Color. 236 233 216)
    :panel-bg-light (Color. 255 255 255)
    :border         (Color. 172 168 153)
    :text-normal    (Color. 0 0 0)
    :text-bright    (Color. 0 0 0)
    :text-light     (Color. 80 80 80)
    :text-dim       (Color. 150 150 150)
    :accent-orange  (Color. 0 85 225)
    :accent-blue    (Color. 49 106 197)
    :accent-green   (Color. 58 110 165)
    :track-header   (Color. 192 192 192)
    :clip-midi      (Color. 255 100 100)
    :clip-audio     (Color. 100 100 255)
    :clip-playing   (Color. 100 255 100)}})

;; ---------------------------------------------------------------------------
;; Mutable theme state (atom)
;; ---------------------------------------------------------------------------

(defonce ^:private state
  (atom {:preset     :ableton-dark
         :scaling    1.0
         :font-size  11
         :font-family "SansSerif"
         :listeners  []}))

(defn- make-fonts [{:keys [font-family font-size scaling]}]
  (let [sz (int (* font-size scaling))]
    {:font-ui       (Font. font-family Font/PLAIN sz)
     :font-ui-bold  (Font. font-family Font/BOLD sz)
     :font-display  (Font. "Monospaced" Font/BOLD (int (* 14 scaling)))}))

(defn theme
  "Returns current theme map: colors merged with fonts and config."
  []
  (let [s @state]
    (merge (get presets (:preset s))
           (make-fonts s)
           (select-keys s [:scaling :font-size :font-family :preset]))))

(defn color
  "Shorthand: (color :bg-dark) => current Color."
  [k]
  (get (theme) k))

(defn font [k] (get (theme) k))

(defn scale
  "Scale an int or float by the current scaling factor."
  [v]
  (let [s (:scaling @state)]
    (if (float? v) (* v s) (int (* v s)))))

(defn update-theme!
  "Update preset, scaling, font-size, and/or font-family. Notifies listeners."
  [& {:keys [preset scaling font-size font-family]}]
  (swap! state (fn [s]
                 (cond-> s
                   preset      (assoc :preset preset)
                   scaling     (assoc :scaling scaling)
                   font-size   (assoc :font-size font-size)
                   font-family (assoc :font-family font-family))))
  (doseq [f (:listeners @state)] (f)))

(defn add-listener! [f]
  (swap! state update :listeners conj f))

;; ---------------------------------------------------------------------------
;; Grid mode — an enum-like keyword set with interval helpers
;; Ported from GridMode.java
;; ---------------------------------------------------------------------------

(def grid-modes
  [:auto :seconds :bar :half :quarter :eighth :sixteenth :thirty-second
   :triplet-quarter :triplet-eighth :triplet-16th :triplet-32nd])

(defn tick-interval
  "Get tick interval for a fixed grid mode given resolution (ticks/beat)."
  [mode resolution]
  (let [bar (* resolution 4)]
    (case mode
      :seconds  resolution
      :bar      bar
      :half     (/ bar 2)
      :quarter  resolution
      :eighth   (/ resolution 2)
      :sixteenth (/ resolution 4)
      :thirty-second (/ resolution 8)
      :triplet-quarter (/ bar 3)
      :triplet-eighth  (/ bar 6)
      :triplet-16th    (/ bar 12)
      :triplet-32nd    (/ bar 24)
      resolution)))

(defn auto-tick-interval
  "Auto-select finest tick interval that keeps >= min-px pixel spacing."
  [resolution tick-width min-px]
  (let [bar (* resolution 4)
        divisions [(/ resolution 8) (/ resolution 4) (/ resolution 2)
                   resolution (/ bar 2) bar]]
    (or (first (filter #(>= (* % tick-width) min-px) divisions))
        bar)))

(defn seconds-interval
  "Get seconds interval for a fixed grid mode given seconds-per-beat."
  [mode seconds-per-beat]
  (let [spb seconds-per-beat
        bar (* spb 4)]
    (case mode
      :seconds  1.0
      :bar      bar
      :half     (/ bar 2)
      :quarter  spb
      :eighth   (/ spb 2)
      :sixteenth (/ spb 4)
      :thirty-second (/ spb 8)
      :triplet-quarter (/ bar 3)
      :triplet-eighth  (/ bar 6)
      :triplet-16th    (/ bar 12)
      :triplet-32nd    (/ bar 24)
      spb)))

(defn auto-seconds-interval
  "Auto-select finest seconds interval that keeps >= min-px pixel spacing."
  [seconds-per-beat px-per-sec min-px]
  (let [bar (* seconds-per-beat 4)
        divisions [(/ seconds-per-beat 8) (/ seconds-per-beat 4)
                   (/ seconds-per-beat 2) seconds-per-beat
                   (/ bar 2) bar]]
    (or (first (filter #(>= (* % px-per-sec) min-px) divisions))
        bar)))

(defn black-key?
  "True if MIDI pitch is a black key."
  [pitch]
  (let [n (mod pitch 12)]
    (contains? #{1 3 6 8 10} n)))
